enhancements/art/flexibly-enforcing-product-build-constraints-in-ci.md

```diff
@@ -0,0 +1,276 @@
+8. **Day N**: If, at any point after feature freeze, the ART build pipeline detects that an upstream repository is not using 
+   the builder image defined in ART metadata, the release will be marked as inconsistent (visible in the release controller 
+   and programmatically accessible by checking an annotation). If the ART pipeline attempts to promote an inconsistent release, 
+   it will abort unless it is manually overridden. The manual override would only be undertaken if it was determined, 
+   with architectural approval, that the inconsistency was introduced erroneously and that the promotion is a critical 
+   delivery (e.g. a CVE fix).
+   
+### Implementation Details/Notes/Constraints
+
+If an upstream component owner does not just accept PRs (e.g. selects an arbitrary go version or selects 
+an out of support ART equivalent from several releases ago), ART will not be able to reverse lookup the product 
+equivalent image. The component owner may believe they are building using a particular version of golang in 
+production, while they are not. If a reverse lookup fails, ART will build with the component using
+the current state of ART metadata. It should be noted that, during this time, an alignment PR will remain 
+open and the overall release will be annotated as inconsistent (details follow).
+
+The proposal also implies that CI will make use of the FROM lines in the repository Dockerfiles. CI can 
+and will apply a mapping to these, but will derive the Golang version (or other versions of that matter) 
+from the repository files. The case of explicit image overrides in the openshift/release repository will become the exception, 
+i.e. a PR against the upstream repository will be enough to switch the version of the base or build image.
+
+When the ART pipeline detects that a given build has inconsistent upstream and downstream configurations, 
+it will still feed the build into the OpenShift release controller. However, the imagestream tag within 
+the 4.x-art-latest imagestream (which the release controller consumes) will be annotated with the reason 
+ART believes the subsequent release cannot be used for a production. The release controller will propagate 
+each inconsistency annotation within the ART imagestream into the subsequent release payload's imagestream 
+tag in is/release. 
+
+For example, if:
+`istag/4.5-art-latest:tests` has the annotation `"release.openshift.io/inconsistency": [ "Upstream and downstream build configurations differ" ]`
+`istag/4.5-art-latest:machine-os-content` has the annotation `"release.openshift.io/inconsistency": [ "Image build with disparate rpms" ]`
+
+The subsequent entry for a release in is/release would be a union of all istag inconsistency annotations:
+`istag/release:4.5.0...` would have the annotation `"release.openshift.io/inconsistency": [ "Image build with disparate rpms", "Upstream and downstream build configurations differ" ]`
+
```

Comment:
I am concerned about the overall size of the `release` imagestream if/when there is a massive influx of imagestreamtags with the `release.openshift.io/inconsistency` annotation.  There is approximately a 2MB limit on the object size that `etcd` will allow.  The current size of the release imagestream is:

$ oc -n ocp get is release -o json | jq -c | wc -c
1217058

Do we have any idea what the maximum size of the "union of all istag inconsistency annotations" might be?

Reply:
I would expect a maximum of two strings here -- so perhaps 160 bytes. If that is a deal breaker, would could instead use key values and a configmap to accomplish something similar. e.g.

release.openshift.io/inconsistency = [ 'a', 'b' ]

and within the configmap, some yaml like:
```yaml
a: Image build with disparate rpms
b: Upstream and downstream build configurations differ
```

Reply:
OK.  I think that's reasonable.  As long as there are only a couple unique messages and they are not excessively verbose, then using the annotation should be acceptable. 

My fear would be that folks using the popup, to see why a release has been flagged as inconsistent, will actually want to see which specific components were flagged as well.  This type of mapping would not be OK to store in the annotation.

Reply:
Updated with: https://github.com/openshift/enhancements/pull/484/commits/e8dc11e83832b0b14aec7e6378fe1bea721ca3ef